Add Compact Ladder Command

sleepywitch opened the pull request:
Adds a compact ladder, a ladder tier, and a rivals ladder command/
I'm open to suggestions for better naming as well as if we want to change the ladder tier sizes or captain images or whatever.
Also, any efficiency or code style comments are helpful since i kinda just went for it with this code.

sleepywitch committed:
Adds Ladder Commands

Adds a compact ladder, a ladder tier, and a rivals ladder command

Pokebunny commented on main/cogs/ladder.py:
```diff
@@ -1,14 +1,37 @@
+# tier_captains should be the same length as tier_min_elo
+tier_captains = [Char.BOWSER, Char.DK, Char.YOSHI, Char.BIRDO, Char.LUIGI, Char.MARIO, Char.PEACH, Char.DAISY, Char.DIDDY, Char.WALUIGI, Char.WARIO]
+tier_min_elo = [2000, 1800, 1600, 1400, 1200, 1000, 800, 600, 400, 200, 0]
+# tier_min_elo 's last number should be 0
```
These should probably be defined as a list of tuples, a dict, or some other data structure that directly ties the two lists together. Also think we could have the tiers more densely clustered at ratings that active players commonly achieve (1000-2000) for better progression opportunity in those tiers at the cost of removing a few tiers from the bottom

Pokebunny commented on main/cogs/ladder.py:
````diff
@@ -40,7 +63,184 @@ async def ladder(self, ctx, mode="off"):
+                        line_count += 1
+                    tier_message += "```\n"
+                    line_count += 2
+                if line_count > 40:
````
Doesn't really look like it but it would be nice if there was a way to have slightly less gap between the bottom of each "code block" and the line that begins the next tier

Pokebunny commented on main/cogs/ladder.py:
```diff
@@ -40,7 +63,184 @@ async def ladder(self, ctx, mode="off"):
+
+    @commands.command(name="rivalsLadder", help="Display the ladder of player and their opponents for their most recent X games (default 10)")
+    @commands.cooldown(1, 300, commands.BucketType.user)
+    async def rivals_ladder(self, ctx, user, mode="off", num_games=10):
```
I think the !rivalsLadder function (or the spirit of it) could be reworked to display the players you have played against the most in the current season or something
A random assortment of your last x games seems less interesting to me than "this is how I compare to people who I've seen a lot lately/this season and likely formed some kind of thoughts about"

sleepywitch replied:
Agreed.

Pokebunny commented on main/cogs/ladder.py:
```diff
@@ -40,7 +63,184 @@ async def ladder(self, ctx, mode="off"):
+
+    @commands.command(name="compactLadder", help="Display the ladder in a compact view [mode] [min]")
+    @commands.cooldown(1, 60, commands.BucketType.default)
+    async def compact_ladder(self, ctx, mode="off", min_games=0):
```
My maybe hot take would be that I would rather have the ladder be a couple characters misaligned for SuleimantheLawgiver than make all output wider just for the very rare extremely long username
Compact is basically just the same ladder but with win % removed. So that makes me think is there something else we could do to adjust the spacing of the main ladder display and make it unnecessary? Remove a decimal place from win %? Adjust something else about the display? Etc
OR - are there more things we could ADD to the default ladder command to make it more interesting/meaningful to people who want a more full display
That might be the better approach to keep both commands meaningful and give people the ladder display they are looking for

sleepywitch replied:
yeah, this is a good thought. gonna combine it with what was discussed in the discord about capping the length for usernames.

sleepywitch committed:
LADDER COMPACT ONLY

Updates the ladder changes to have ladder compact only since that's the most important one from a function standpoint

sleepywitch committed:
Update ladder.py

Removing unnecessary variables and functions

sleepywitch commented:
Updated to just add the !ladderCompact command
